pydrive2/auth.py

```diff
@@ -226,120 +213,123 @@ def LocalWebserverAuth(
-            # values, as each call's changes are global.
-        success = False
-        port_number = 0
+            port_numbers = [8080, 8090]
+
+        additional_config = {}
+        # offline token request needed to obtain refresh token
+        # make sure that consent is requested
+        if self.settings.get("get_refresh_token"):
+            additional_config["access_type"] = "offline"
+            additional_config["prompt"] = "select_account"
+
         for port in port_numbers:
-            port_number = port
             try:
-                httpd = ClientRedirectServer(
-                    (host_name, port), ClientRedirectHandler
+                self._credentials = self.flow.run_local_server(
+                    host=host_name,
+                    port=port,
+                    authorization_prompt_message=_CLIENT_AUTH_PROMPT_MESSAGE,
+                    open_browser=launch_browser,
+                    **additional_config,
                 )
-            except OSError:
-                pass
-            else:
-                success = True
-                break
-        if success:
-            oauth_callback = f"http://{host_name}:{port_number}/"
-        else:
-            print(
-                "Failed to start a local web server. Please check your firewall"
-            )
-            print(
-                "settings and locally running programs that may be blocking or"
-            )
-            print("using configured ports. Default ports are 8080 and 8090.")
-            raise AuthenticationError()
-        self.flow.redirect_uri = oauth_callback
-        authorize_url = self.GetAuthUrl()
-        if launch_browser:
-            webbrowser.open(authorize_url, new=1, autoraise=True)
-            print("Your browser has been opened to visit:")
-        else:
-            print("Open your browser to visit:")
-        print()
-        print("    " + authorize_url)
-        print()
-        httpd.handle_request()
-        if "error" in httpd.query_params:
-            print("Authentication request was rejected")
-            raise AuthenticationRejected("User rejected authentication")
-        if "code" in httpd.query_params:
-            return httpd.query_params["code"]
-        else:
-            print(
-                'Failed to find "code" in the query parameters of the redirect.'
-            )
-            print("Try command-line authentication")
-            raise AuthenticationError("No code found in redirect")
 
-    @CheckAuth
+            except OSError as e:
+                # Google Auth's local server implementation does not allow socket reuse
+                # i.e. WSGIServer.allow_reuse_address = False
+                # The Installed App Flow raises OSError when binding to a used port
+
+                # More on TCP state changes: https://users.cs.northwestern.edu/~agupta/cs340/project2/TCPIP_State_Transition_Diagram.pdf
+                if e.errno not in (10048, 48):
```

Comment:
Hmm, I still don't understand this logic - my understanding is that we can just try another port in case of an error (any OSError). 

```
except OSError:
                pass
```

Reply:
Ohhhh yeahh. There's no need for me to specifically catch these 2 error codes. Will update this

Reply:
👍